libgadget/fdm.c

```diff
@@ -0,0 +1,585 @@
+        }
+#endif
+        if(tw->Niteration > FDMMAXITER) {
+            endrun(1155, "failed to converge in neighbour iteration in density()\n");
+        }
+    } while(1);
+
+    ta_free(priv->NPRedo);
+    ta_free(priv->NPLeft);
+    
+    myfree(priv->DhsmlDensity);
+    myfree(priv->NumNgb);
+    myfree(priv->Right);
+    myfree(priv->Left);
+
+    double timeall = walltime_measure(WALLTIME_IGNORE);
+
+    double timecomp = tw->timecomp3 + tw->timecomp1 + tw->timecomp2;
+    double timewait = tw->timewait1 + tw->timewait2;
+    double timecomm = tw->timecommsumm1 + tw->timecommsumm2;
+    walltime_add("/FDM/Density/Compute", timecomp);
+    walltime_add("/FDM/Density/Wait", timewait);
+    walltime_add("/FDM/Density/Comm", timecomm);
+    walltime_add("/FDM/Density/Misc", timeall - (timecomp + timewait + timecomm));
+
+    return;
+}
+
+/***************************************************************************************************/
+/* Now we start to do the QP iterations*/
+
+typedef struct{
+    TreeWalkQueryBase base;
+    MyFloat Hsml;
+    MyFloat Density;
+} TreeWalkQueryDerivative;
+
+typedef struct{
+    TreeWalkResultBase base;
+    MyFloat GradDensity[3];
+    MyFloat LapDensity;
+} TreeWalkResultDerivative;
+
+typedef struct {
+    TreeWalkNgbIterBase base;
+    DensityKernel kernel_i;
+} TreeWalkNgbIterDerivative;
+
+
+static void
+densderiv_copy(int place, TreeWalkQueryDerivative * I, TreeWalk * tw)
+{
+    I->Hsml = P[place].Hsml;
+    I->Density = FDMP(place).Density;
+}
+
+static void
+densderiv_reduce(int place, TreeWalkResultDerivative * remote, enum TreeWalkReduceMode mode, TreeWalk * tw)
+{
+    TREEWALK_REDUCE(FDMP(place).LapDensity, remote->LapDensity);
+    int k;
+    for(k=0;k<3;k++){
+        TREEWALK_REDUCE(FDMP(place).GradDensity[k], remote->GradDensity[k]);
+    }
+}
+
+static void
+densderiv_ngbiter(
+        TreeWalkQueryDerivative * I,
+        TreeWalkResultDerivative * O,
+        TreeWalkNgbIterDerivative * iter,
+        LocalTreeWalk * lv)
+{
+    if(iter->base.other == -1) {
+        iter->base.Hsml = I->Hsml;
+        iter->base.mask = 2; /* dm-only */
+        iter->base.symmetric = NGB_TREEFIND_ASYMMETRIC;
+        
+        density_kernel_init(&iter->kernel_i, I->Hsml, GetDensityKernelType());
+        return;
+    }
+    
+    const int other = iter->base.other;
+    const double r = iter->base.r;
+    const double r2 = iter->base.r2;
+    const double * dist = iter->base.dist;
+    
+    if(P[other].Type != 1 || r2 <=0 || !(r2 < iter->kernel_i.HH))
+        return;
+    
+    const double u = r * iter->kernel_i.Hinv;
+    const double dwk = density_kernel_dwk(&iter->kernel_i, u);
+    const double ddwk = density_kernel_ddwk(&iter->kernel_i, u);
+    const double mass_j = P[other].Mass;
+    
+    double fac = (FDMP(other).Density - I->Density)/sqrt(I->Density*FDMP(other).Density);
+    /* \Laplacian(W(r,h)) = 2*dwk/r + ddwk */
```

Comment:
Check that I did the laplacian correct:
q = r/h; d/dr = (1/h)d/dq
=> d/dx(W(r,h)) = dr/dx * (1/h)dW/dq = x/r * dwk
=> d^2/dx^2(W(r,h)) = d/dx(x/r)*dwk + (x^2/r^2)*(1/h^2)d^2W/dq^2 = (1/r-x^2/r^3)*dwk + (x^2/r^2)*ddwk

Reply:
This should be correct. On idealized test of 1D hyperbolic field, the second order derivative matches the analytical result well. 
![hyperbolic1D](https://user-images.githubusercontent.com/52514409/113163224-f48c5600-920d-11eb-880e-dfaea0b346d1.png)